### PR TITLE
Implement equals/hashcode on Rule and subfields

### PR DESCRIPTION
**Pull request by AZWN**
This PR implements `hashCode` and `equals` on all (transitive) subfields of `Rule`.

### CODE REVIEW

File: nabl2.terms/src/main/java/mb/nabl2/regexp/RegExpMatcher.java [hunk 1]
```diff
@@ -136,4 +137,20 @@ private RegExpMatcher(IRegExp<S> state, Map<IRegExp<S>, Map<S, IRegExp<S>>> stat
         return new RegExpMatcher<>(initial, stateTransitions, defaultTransitions, nonFinal, isNullable);
     }
 
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        RegExpMatcher<?> that = (RegExpMatcher<?>) o;
+        return Objects.equals(state, that.state) &&
```
Comment (hendrikvanantwerpen) [inline nabl2.terms/src/main/java/mb/nabl2/regexp/RegExpMatcher.java:145]: It is enough to only compare `state` here. All other fields are derived from that regular expression.

File: nabl2.terms/src/main/java/mb/nabl2/regexp/RegExpMatcher.java [hunk 2]
```diff
@@ -136,4 +137,20 @@ private RegExpMatcher(IRegExp<S> state, Map<IRegExp<S>, Map<S, IRegExp<S>>> stat
...
+                Objects.equals(stateTransitions, that.stateTransitions) &&
+                Objects.equals(defaultTransitions, that.defaultTransitions) &&
+                Objects.equals(nonFinal, that.nonFinal) &&
+                Objects.equals(isNullable, that.isNullable);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(state, stateTransitions, defaultTransitions, nonFinal, isNullable);
```
Comment (hendrikvanantwerpen) [inline nabl2.terms/src/main/java/mb/nabl2/regexp/RegExpMatcher.java:154]: Same here, only `state` is enough.

**Comment by AZWN** [comment]
Implemented your suggestions. 2 questions/remarks:
- The `equals`/`hashCode` for `Relation.Immutable` was [already implemented](https://github.com/metaborg/nabl/blob/7d237186daf6609d53d78526af30ddf60689a855/nabl2.terms/src/main/java/mb/nabl2/relations/impl/Relation.java#L212-L233), right?
- I am curious how I missed `Relation.Transient`, because I can only find `Relation.Immutable` referenced via (`CResolveQuery.min -> `IQueryMin.labelOrd`)?

**Comment by hendrikvanantwerpen** [comment]
Thanks, looks good. I overlooked that `Relation` already implemented `equals`. And it could be that `Relation.Transient` is never referenced via `Rule` directly. But it's good to have it on both versions of the data structure, I guess.

**Pull request merged**